Source/WebCore/platform/graphics/gstreamer/eme/CDMOpenCDM.cpp

```diff
@@ -26,16 +26,117 @@
+    void addObserver(OpenCDMSession*, Observer*);
+    void removeObserver(Observer*);
+
+private:
+    using Notification = void (Observer::*)(String&&);
+    OCDMSessionCallbacksNotifier()
+    {
+        m_callbacks.process_challenge = [](OpenCDMSession* session, const char[], const uint8_t challenge[], const uint16_t challengeLength) {
+            OCDMSessionCallbacksNotifier::notify(session, &Observer::challengeGeneratedCallback, challenge, challengeLength);
+        };
+        m_callbacks.key_update = [](OpenCDMSession* session, const uint8_t key[], const uint8_t keyLength) {
+            OCDMSessionCallbacksNotifier::notify(session, &Observer::keyUpdatedCallback, key, keyLength);
+        };
+        m_callbacks.message = [](OpenCDMSession* session, const char message[]) {
+            OCDMSessionCallbacksNotifier::notify(session, &Observer::messageReceivedCallback, reinterpret_cast<const uint8_t*>(message), strlen(message));
+        };
+    }
+
+    static bool tryNotify(OpenCDMSession*, Notification, Ref<WebCore::SharedBuffer>&&);
+    static void notify(OpenCDMSession*, Notification, const uint8_t[], uint16_t);
+
+    friend class NeverDestroyed<OCDMSessionCallbacksNotifier>;
+    OpenCDMSessionCallbacks m_callbacks;
+    HashMap<OpenCDMSession*, Observer*> m_observers;
+    WebCore::GenericTaskQueue<WebCore::Timer> m_taskQueue;
+};
+
+OCDMSessionCallbacksNotifier& OCDMSessionCallbacksNotifier::singleton()
+{
+    static NeverDestroyed<OCDMSessionCallbacksNotifier> instance;
+    return instance;
+}
+
+void OCDMSessionCallbacksNotifier::addObserver(OpenCDMSession* session, Observer* observer)
+{
+    m_observers.set(session, observer);
+}
+
+void OCDMSessionCallbacksNotifier::removeObserver(Observer* observer)
+{
+    m_observers.removeIf([observer](const auto& keyAndValue) { return observer == keyAndValue.value; });
+}
+
+bool OCDMSessionCallbacksNotifier::tryNotify(OpenCDMSession* session, Notification method, Ref<WebCore::SharedBuffer>&& message)
+{
+    if (!isMainThread()) {
+        // Make sure all happens on the main thread to avoid locking.
+        callOnMainThread([session, method, message = WTFMove(message)]() mutable {
+            OCDMSessionCallbacksNotifier::tryNotify(session, method, WTFMove(message));
+        });
+        return true;
+    }
+
+    OCDMSessionCallbacksNotifier& self = OCDMSessionCallbacksNotifier::singleton();
+    auto entry = self.m_observers.find(session);
+    if (entry != self.m_observers.end()) {
+        (entry->value->*method)(String(message->data(), message->size()));
+        return true;
+    }
+
+    return false;
+}
+
+void OCDMSessionCallbacksNotifier::notify(OpenCDMSession* session, Notification method, const uint8_t message[], uint16_t messageLength)
+{
+    auto sharedMessage = WebCore::SharedBuffer::create(message, messageLength);
+    if (OCDMSessionCallbacksNotifier::tryNotify(session, method, WTFMove(sharedMessage)))
```

Comment:
if this retuns false the` sharedMessage` below is no longer valid. It go moved.

Reply:
use `sharedMessage->copy()` here

Reply:
Nope: https://github.com/WebPlatformForEmbedded/WPEWebKit/pull/583#discussion_r271741670